include/oneapi/dpl/pstl/execution_defs.h

```diff
@@ -123,27 +123,30 @@ inline constexpr unsequenced_policy unseq{};
 
 // 2.3, Execution policy type trait
 template <class T>
-struct is_execution_policy : ::std::false_type
+struct __is_execution_policy_impl : ::std::false_type
 {
 };
 
 template <>
-struct is_execution_policy<oneapi::dpl::execution::sequenced_policy> : ::std::true_type
+struct __is_execution_policy_impl<oneapi::dpl::execution::sequenced_policy> : ::std::true_type
 {
 };
 template <>
-struct is_execution_policy<oneapi::dpl::execution::parallel_policy> : ::std::true_type
+struct __is_execution_policy_impl<oneapi::dpl::execution::parallel_policy> : ::std::true_type
 {
 };
 template <>
-struct is_execution_policy<oneapi::dpl::execution::parallel_unsequenced_policy> : ::std::true_type
+struct __is_execution_policy_impl<oneapi::dpl::execution::parallel_unsequenced_policy> : ::std::true_type
 {
 };
 template <>
-struct is_execution_policy<oneapi::dpl::execution::unsequenced_policy> : ::std::true_type
+struct __is_execution_policy_impl<oneapi::dpl::execution::unsequenced_policy> : ::std::true_type
 {
 };
 
+template <typename T>
+using is_execution_policy = __is_execution_policy_impl<::std::decay_t<T>>;
```

Comment:
After our offline discussion, I think `is_execution_policy` should not decay the type (as it changes the semantics of a public and defined struct). We should add a layer which we use internally to decay the type automatically if we want that functionality.

Direct usages of this public interface should go back to decaying the type explicitly (in the tests for instance).

The other (internal) versions of this can probably stay as they are.

Reply:
> After our offline discussion, I think `is_execution_policy` should not decay the type (as it changes the semantics of a public and defined struct). We should add a layer which we use internally to decay the type automatically if we want that functionality.
> 
> Direct usages of this public interface should go back to decaying the type explicitly (in the tests for instance).
> 
> The other (internal) versions of this can probably stay as they are currently in the PR.

@danhoeflinger, thanks for explaining. That is exactly what I had in my mind to propose about this issue:)

Reply:
Good to hear we are on the same page.  

The only thing I am unsure of is if we want our internal versions of this pattern to mirror the semantics of this public one.   
I think it is probably unnecessary and will just complicate code and make names longer, but its something worth considering.